### ISSUE TITLE
Add support & documentation on using Azure CNI with capz

### ISSUE THREAD

**Issue by CecileRobertMichon** [kubernetes-sigs/cluster-api-provider-azure#467]
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
https://github.com/Azure/azure-container-networking/blob/master/docs/cni.md

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

- [ ] https://github.com/Azure/azure-container-networking/issues/1945
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3343
- [x] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3342
- [x] https://github.com/Azure/azure-container-networking/issues/1897

A little lower priority issue, but must be solved to support multiple NICs on control plane.
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3506

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by richardcase** [comment]
This would be good for installing the CNI: https://github.com/kubernetes-sigs/cluster-api/pull/3050

**Comment by CecileRobertMichon** [comment]
@mboersma is this one you'd be interested in working on? It's a relatively hi-pri one because it's a blocker for IPv6 (cc @jsturtevant)

**Comment by mboersma** [comment]
@CecileRobertMichon yes indeed, I can probably start on this tomorrow.

**Comment by jackfrancis** [comment]
An implementation question: is the version of Azure CNI currently configurable?

cc @craiglpeters

**Comment by CecileRobertMichon** [comment]
@jackfrancis not sure I understand the question? There is no Azure CNI implementation currently.

**Comment by jackfrancis** [comment]
A reminder to use the "transparent" mode configuration when implementing a capz + Azure CNI scenario. This AKS Engine PR configures Azure CNI for "transparent" mode by default:

https://github.com/Azure/aks-engine/pull/3958

The vanilla installation method that AKS Engine follows (download the public release tarball and untar/gzip it) delivers a default configuration of "bridge" mode. AKS Engine modifies this to "transparent" by sed replacing the appropriate value:

https://github.com/Azure/aks-engine/blob/master/parts/k8s/cloud-init/artifacts/cse_config.sh#L288

Not pretty, but there you have it.

**Comment by CecileRobertMichon** [comment]
ref https://github.com/Azure/azure-container-networking/blob/master/acncli/deployment/manager.yaml

**Comment by jsturtevant** [comment]
I think this is still something we would want to look at eventually

/remove-lifecycle stale

**Comment by CecileRobertMichon** [comment]
To summarize the current blockers here:

- Azure CNI install is via binary. There is no helm chart, operator or daemonset which would allow users to install the CNI similarly to other CNIs after the cluster is created, unlike other CNIs currently documented with CAPZ.
- Azure CNI requires static pre-allocation of IPs. This means that CAPZ must be aware of CNI to pre-allocate the IPs (this is unusual as CNIs are usually installed on the cluster independently of infra and infrastructure provider has no knowledge of CNI). It also means it must use the maxPods setting which is defined in kubelet settings (not currently part of the bootstrap provider [contract](https://cluster-api.sigs.k8s.io/developer/providers/bootstrap.html?highlight=sentin#bootstrap-api-resource)), and that the IP allocation must be dynamic if the max pods value is updated.

**Comment by ravidbro** [comment]
@CecileRobertMichon about the second bullet, isn't it the same on any other cloud provider like EKS and GKE?
Don't CAPI needs to be aware there to the platform too?

**Comment by CecileRobertMichon** [comment]
@ravidbro not sure I understand your question. EKS and GKE are both managed Kubernetes services. This issue is specific to self-managed Azure clusters (not AKS). CAPI supports a "BYO CNI" model for self-managed clusters, which means the user can install the CNI independently of the cluster definition and there isn't anything in the Cluster or InfraCluster spec that is CNI aware.

**Comment by ravidbro** [comment]
I understand, thanks!

**Comment by dmeytin** [comment]
Any progress with two blockers above ?

**Comment by CecileRobertMichon** [comment]
@matmerr @tamilmani1989 any updates from the Azure CNI side? Blocker are listed [above](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/467#issuecomment-1039396096).

**Comment by tamilmani1989** [comment]
Couldn't add evan here.. he s working on installing cni via cns daemonset. Let me notify him

**Comment by rbtr** [comment]
👋🏼
I'm working on an azcni installer that we can plug in to AKS BYO CNI, that will satisfy blocker #1

**Comment by dthorsen** [comment]
@rbtr @CecileRobertMichon I was just experimenting with the Azure CNI in a self-managed cluster and I used [this daemonset](https://github.com/Azure/azure-container-networking/blob/57d6d636a67812e6aac447f608f908cb6d624123/tools/acncli/deployment/manager.yaml). It seemed to work well, with the exception that I needed a separate network interface in a second subnet for pods, and of course ipconfigs on the second network interface. We were thinking of PR'ing the AzureMachine stuff to add support for multiple network interfaces and ipconfigs. (Just a static number). Then I think we could just set max pods in the kubelet via a kubelet arg?

**Comment by CecileRobertMichon** [comment]
>  We were thinking of PR'ing the AzureMachine stuff to add support for multiple network interfaces and ipconfigs. (Just a static number)

I like the idea of making this generic, and not specific to Azure CNI. CAPZ should not know/care about CNIs. This might also help with #2150 down the road. 

> Then I think we could just set max pods in the kubelet via a kubelet arg?

Can you please expand on that? are you saying the user would have to make sure the ipconfig number matches max pods or would capz somehow dynamically get that number from kubelet? That's been the biggest antipattern with this issue so curious to know how you're approaching it.

**Comment by dthorsen** [comment]
> are you saying the user would have to make sure the ipconfig number matches max pods or would capz somehow dynamically get that number from kubelet?

Yes, I was considering manually making sure that I set the max pods the same as the ipconfig count. I mean, dynamic would be even better, but if I understand the current contract situation correctly, that seems impossible without modification of the contract. I imagine that will be more complex and have more lead time, so I was just looking to try a simple workaround by statically configuring the max pods to equal the number of ipconfigs I attach per node.

**Comment by CecileRobertMichon** [comment]
Agreed. I think we would need the [2nd bullet point above](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/467#issuecomment-1039396096) in order to make it dynamic

**Comment by rbtr** [comment]
We're a way off from shipping this in AKS still, but a [CNI installer container](https://github.com/Azure/azure-container-networking/blob/master/test/integration/manifests/cns/daemonset.yaml#L76-L92) is being used to run AzCNI e2e tests in our pipelines now 🙂 
It's only set up for the Swift (dynamic Pod IP) scenario currently but is flexible. For the CAPZ ask, is it sufficient that it supports the AzCNI v1/static Pod IP style deployments?

**Comment by dthorsen** [comment]
@nawazkh I am cleaning up [this PR](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2411) as we speak. Working out a bug before I push.

**Comment by CecileRobertMichon** [comment]
@dthorsen I started looking at this with @nawazkh a few weeks ago as we have an internal customer who is blocked on this, how close are you to get it ready for review again? Want to make sure we're not duplicating effort here but we have a tight deadline on this one

**Comment by dthorsen** [comment]
Ah! I see! I intend to get what I have pushed up today. I believe I addressed nearly all feedback. A rebase will depend a bit on the number of conflicts, but probably could have it done by end of day tomorrow.

**Comment by CecileRobertMichon** [comment]
@dthorsen that's great, if that's the case we'll help review and validate the PR for https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2327 and @nawazkh and I can follow up with an e2e test + doc on using Azure CNI to close out this issue

**Comment by dthorsen** [comment]
@CecileRobertMichon @nawazkh I have pushed my commit. The tests are passing and I spun up a few workload clusters from tilt with various privateIPConfiguration counts, public IP allocation enabled/disabled, etc. I have to sign off for the night but let me know what you think tomorrow.

**Comment by dthorsen** [comment]
Here is a link to the Azure CNI Manager daemonset that can be used to install the Azure CNI quite easily.

https://github.com/Azure/azure-container-networking/blob/ef1bff6046156dec56b46664f516921bd423e9e0/tools/acncli/deployment/manager.yaml

**Comment by rbtr** [comment]
@dthorsen this is not a supported tool, don't recommend it for this use please.

**Comment by dthorsen** [comment]
That is disappointing, because it seems to do exactly what is required. It was very easy to get up and running this way. Is there a way it could become a supported tool? Otherwise it may be something that makes sense for CAPZ to fork and support, but that seems somewhat redundant if there is already a tool that does this. The functionality needed is pretty trivial, basically just install the CNI binaries, drop the config, and sleep forever.

**Comment by jackfrancis** [comment]
@rbtr @tamilmani1989 can you provide more information? In lieu of an official solution, why is wrapping a daemonset around the published manifest in the official Azure CNI repo not recommended?

**Comment by rbtr** [comment]
@jackfrancis @dthorsen in short, the image referenced by that manifest is unmaintained and not prod-ready. 

We built it for our CI/nonprod debugging with no intent of public use. It is technically capable of installing the CNI binary, but I can't support CAPZ using/recommending it for use. Maybe most importantly, we have moved past needing it internally and it isn't getting updates.

I did [previously try to engage here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/467#issuecomment-1209918061) to offer what will be an official supported solution but that didn't seem to get anyone's attention. [This tool is being used for CNI install in AKS](https://github.com/Azure/azure-container-networking/blob/ef1bff6046156dec56b46664f516921bd423e9e0/hack/manifests/cni-installer.yaml) already and could probably trivially be extended for CAPZ, but I need some clarification on the requirements to make that happen. I suspect [this](https://github.com/Azure/azure-container-networking/pull/1701) will be sufficient, but if some CAPZ folks would like to contribute, we accept PRs 🙂

**Comment by jackfrancis** [comment]
It seems based on @dthorsen's experience that the now-abandoned installer image can be used as the set of functional requirements. @dthorsen did that container image simply install the ipam stuff with a sufficient amount of UX configuration to fulfill all of the needs of Azure CNI v1 on a capz-built node? E.g.:

```
env:
            - name: AZURE_CNI_OS
              value: <insert value here>
            - name: AZURE_CNI_TENANCY
              value: <insert value here>
            - name: AZURE_CNI_MODE
              value: <insert value here>
            - name: AZURE_CNI_IPAM
              value: <insert value here>
            - name: AZURE_CNI_EXEMPT
              value: <insert value here>
```

@rbtr would it be possible to extend the tool AKS is using to (1) include v1 of the CNI components with (2) the configuration above?

**Comment by nawazkh** [comment]
Sorry, I haven't been active at posting updates on this test effort but I am actively working on this!
So far, Cecile and I have tested Dane's PR of allocating PrivateIPConfigs to a VM, and it works as expected. 
I now am working on getting Azure CNI up-and-running in a test cluster.

**Comment by nawazkh** [comment]
I created two other issues to track the branched out effort to enable Azure CNI on CAPZ
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3343
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3342

**Comment by nawazkh** [comment]
One of the edge case of "support for configurable network interfaces" discovered while testing 
- https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3506

**Comment by nawazkh** [comment]
closing this issue in favor of epic tracking this effort https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3611
/close
